src/Common/src/TypeSystem/IL/Stubs/StructMarshallingThunk.cs

```diff
@@ -0,0 +1,266 @@
+
+    internal class StructMarshallingThunk : ILStubMethod
+    {
+        internal readonly MetadataType ManagedType;
+        internal readonly NativeStructType NativeType;
+        internal readonly StructMarshallingThunkType ThunkType;
+        private  InteropStateManager _interopStateManager;
+        private TypeDesc _owningType;
+
+        public StructMarshallingThunk(TypeDesc owningType, MetadataType managedType, StructMarshallingThunkType thunkType, InteropStateManager interopStateManager)
+        {
+            _owningType = owningType;
+            ManagedType = managedType;
+            _interopStateManager = interopStateManager;
+            NativeType = _interopStateManager.GetStructMarshallingNativeType(managedType);
+            ThunkType = thunkType;
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return ManagedType.Context;
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get
+            {
+                return _owningType;
+            }
+        }
+
+        private MethodSignature _signature;
+        public override MethodSignature Signature
+        {
+            get
+            {
+                if (_signature == null)
+                {
+                    TypeDesc[] parameters;
+                    if (ThunkType == StructMarshallingThunkType.Cleanup)
+                    {
+                        parameters = new TypeDesc[] {
+                            NativeType.MakeByRefType()
+                        };
+                    }
+                    else
+                    {
+                        parameters = new TypeDesc[] {
+                            ManagedType.MakeByRefType(),
+                            NativeType.MakeByRefType()
+                        };
+                    }
+                    _signature = new MethodSignature(MethodSignatureFlags.Static, 0, Context.GetWellKnownType(WellKnownType.Void), parameters);
+                }
+                return _signature;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                if (ThunkType == StructMarshallingThunkType.ManagedToNative)
+                {
+                    return "ManagedToNative__" + ((MetadataType)ManagedType).Name;
+                }
+                else if (ThunkType == StructMarshallingThunkType.NativeToManage)
+                {
+                    return "NativeToManaged__" + ((MetadataType)ManagedType).Name;
+                }
+                else
+                {
+                    return "Cleanup__" + ((MetadataType)ManagedType).Name;
+                }
+            }
+        }
+
+        private int GetNumberOfInstanceFields()
+        {
+            IEnumerable<FieldDesc> fields = ManagedType.GetFields();
+            int result = 0;
+            using (IEnumerator<FieldDesc> enumerator = fields.GetEnumerator())
```

Comment:
Is there a specific reason not to just use a foreach loop? At first I thought you're trying to avoid the finally block (that disposes the enumerator by doing nothing in our case), but this seems like precisely what a `foreach (FieldDesc field in fields)` would expand into, but it's harder to read.